proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,321 @@
+
+```swift
+protocol P {
+  func foo() -> Self
+  
+  func bar(_: Self)
+}
+```
+
+Accessing a member on an existential value requires the ability to spell the type of that member outside its protocol context. Today, the one means to representing the dynamic `Self` type of an existential value is type erasure — the substitution of `Self` with a representable supertype, like `P`. On the other hand, type erasure is safe to perform only in [covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) position. For example, calling `foo` on a value of type `P` with its covariant `Self` result type-erased to `P` is safe, whereas allowing one to pass a type-erased value to `bar` would expose the opportunity to pass in an argument of non-matching type.
+
+In contrast to requirements, protocol extension members cannot afford to retroactively jeopardize the existential availability of the entire protocol. This is when the second condition shows its true colors, with the restriction forced to take on a more reasonable, on-demand manifestation in the form of a member access failure:
+
+```swift 
+protocol P {}
+extension P {
+  func method(_: Self) {}
+}
+
+func callMethod(p: P) {
+  p.method // error: member 'method' cannot be used on value of protocol type 'P'; use a generic constraint instead
+}
+```
+
+As we hope it became clear, a requirement such as `func bar(_: Self)` or an associated type requirement alone cannot speak for the rest of the interface. Some protocols still have a useful subset of functionality that does not rely on `Self` and associated types whatsoever, or does so in a way that is compatible with existential values, like `func foo() -> Self`. In a refined protocol, some requirements may also happen to exclusively rely on an associated type with a fully concrete, known implementation, and become safe to invoke using a value of the refined protocol type. This brings us to a well-known implementation hole with counterintuitive behavior; in the snippet below, `Animal` is still assumed as having an associated type requirement despite the same-type constraint that effectively predefines a fully concrete implementation for it.
+
+```swift
+protocol Animal: Identifiable where ID == String {}
+extension Animal {
+  var name: String { id }
+}
+```
+
+The current semantic inconsistency also discourages authors from refining their existing protocols with other, useful ones in fear of losing existential qualification.
+
+### Library Evolution
+
+Removing the type-level restriction would mean that adding defaulted requirements to a protocol is always both a binary- and source-compatible change, since it could no longer interfere with existing uses of the protocol.
+
+### Type-erasing Wrappers
```

Comment:
Should the `W` in `Wrappers` be capitalized?

Reply:
Yeah, I suppose we should capitalize the `e` in `erasing`.